### DISCUSSION TITLE
Installation issues

### DISCUSSION THREAD

**Discussion by siggib007** [pretalx/pretalx#1518] [Self-hosting]
Working on setting up my own self hosted Pretalx and working through the install guide at https://docs.pretalx.org/administrator/installation.html. I'm getting a failure at the `pip install --user --upgrade-strategy eager -U pretalx` command. I'm building this on Debian 12. Here is the error I'm getting
`  ERROR: Failed building wheel for reportlab
  Running setup.py clean for reportlab
Failed to build Pillow reportlab
ERROR: Could not build wheels for Pillow, reportlab, which is required to install pyproject.toml-based projects
`
I've tried to research this but can't make head or tails out of this. Can anyonoe assist or provide some pointers?

**Comment by rixx** [comment]
Weird, because pretalx is not a `pyproject.toml` based project. From the error message, I'd assume that you're missing some dependencies for Pillow and/or reportlab. I'm not super familiar with Debian, so I can't really speculate as to that, but you might find something helpful looking for those terms.

(Also, just in case you didn't, make sure to run `pip install -U setuptools wheel pip` first, because outdated build tools can be part of the problem.)

**Reply by GreenReaper** [reply]
I'm running into this too. Here's the issue, at least as far as I understand it:
* The latest release, from March, is [2.3.2](https://github.com/pretalx/pretalx/releases/tag/v2.3.2).
* This [requires ReportLab~=3.5.0](https://github.com/pretalx/pretalx/blob/v2.3.2/src/setup.py#L128)
* [ReportLab 3.5.x accesses an member of PyFrameObject (struct frame)](https://bugzilla.redhat.com/show_bug.cgi?id=2034174#c3) which is [no longer available in Python 3.11](https://docs.python.org/3/whatsnew/3.11.html#pyframeobject-3-11-hiding)
* Debian 12 ships with Python 3.11.2

It'd be nice to get an updated release which addresses this, although if you have a way to just override the ReportLab dependency to allow ~3.6.0, that'd be great too - I'm not really familiar with Python or its installer system, and someone just asked me to get our existing install working again on a system already upgraded to Bookworm. I can try a source install, but that is obviously not ideal.

**Update:** Setting a custom constraint file led to a dependency error. But installing pretalx with --no-deps, then modifying `local/lib/python3.11/site-packages/pretalx-2.3.2.dist-info/METADATA` to require ~=3.6.0 and running `pip install --upgrade-strategy eager -U pretalx -v` within the venv worked as far as "Successfully installed". I needed to install libfreetype-dev on the system as well, or it complained about missing header file ft2build.h (something to do with _rl_accel, _renderPM and/or libart).

After that gunicorn gets further, printing pretalx's header, but then complaining about `ModuleNotFoundError: No module named 'dj_static'` and `ImportError: cannot import name 'formatargspec' from 'inspect' (/usr/lib/python3.11/inspect.py)`. Overall I'm not sure this is going to end up a viable route since other selected packages may not be compatible with 3.11.

**Comment by siggib007** [comment]
Thanks, is there a different distribution I might have better luck with? I saw mention of Ubuntu somewhere in the instructions, would you recommend redoing the installation on Ubuntu?
BTW yes I did do the setuptools wheel line first, twice in fact just to make sure 😎

Siggi Bjarnason (he/him)
https://supergeek.us
https://infosechelp.net
https://www.linkedin.com/in/siggib
https://twitter.com/siggib007


----- Original message -----

**Reply by rixx** [reply]
I would hope that the Ubuntu instructions are up to date, but if you find the equivalents of the Ubuntu dependencies from the docs, you should be good to go, too. Sorry, but without further information, I don't think there's much anybody can do to help.

**Comment by siggib007** [comment]
I have found no specific Ubuntu instructions, maybe the admin guide I'm using is assuming I'm using Ubuntu but I saw no mention of that. If there are some other "Ubuntu instructions" I'm missing I would love a link to them.
I'll rebuild the machine with the latest Ubuntu release and will try again. I'm not a Linux guru so "finding equivalents" or "just finding the missing dependencies" is not viable for me. It is easier to start over from scratch using Ubuntu if that is going to be easier.

FYI I'm self-hosting for financial reasons, not geeky reasons. 

Thank you for your help, I really appreciate it.

Siggi Bjarnason (he/him)
https://supergeek.us
https://infosechelp.net
https://www.linkedin.com/in/siggib
https://twitter.com/siggib007


----- Original message -----

**Reply by rixx** [reply]
Okay, it doesn't say "Ubuntu" but "Ubuntu-like systems", but still: https://docs.pretalx.org/administrator/installation.html#step-3-package-dependencies

Anyway, I can only reiterate that without further details, I can't help. Your best bet is to look up why reportlab and Pillow don't install on your system.

**Comment by siggib007** [comment]
Thank you, I think I'll rebuild the system with the latest Ubuntu and see how that goes. I feel that is going to be my best bet. I'm not confident in my ability to figure out why reportlab and Pillow aren't installing, especially considering I have no clue what those things are. 
I'd recommend putting something at the top of that guide about Ubuntu being recommended. Most would argue that Debian is an "Ubuntu-like system".

Again, thank you so much for your help.

Siggi Bjarnason (he/him)
https://supergeek.us
https://infosechelp.net
https://www.linkedin.com/in/siggib
https://twitter.com/siggib007


----- Original message -----

**Reply by rixx** [reply]
We do not recommend a particular operating system – any hints in the documentation were contributed by pretalx users.

**Reply by siggib007** [reply]
OK redoing this on Ubuntu 22.04.2 LTS worked much better and I think I'm 99% there. Now I'm failing to start the web service with this error:
```
Jul 05 23:04:29 pretalx systemd[6908]: pretalx-web.service: Changing to the requested working directory failed: No such file or directory
Jul 05 23:04:29 pretalx systemd[6908]: pretalx-web.service: Failed at step CHDIR spawning /var/pretalx/.local/bin/gunicorn: No such file or directory
```
Upon further inspection that path is a file, so it makes sense that it can't change directories into a file. What am I doing wrong?
```
pretalx@pretalx:/tmp$ ls -al /var/pretalx/.local/bin/gunicorn
-rwxrwxr-x 1 pretalx pretalx 217 Jul  5 21:53 /var/pretalx/.local/bin/gunicorn
pretalx@pretalx:/tmp$ 
```

**Reply by rixx** [reply]
… Why are you trying to change into the gunicorn file? Please always provide both the error message and the reason it occurs, in this case presumably the contents of your systemd file etc.

**Reply by siggib007** [reply]
I'm not trying anything, but to start the service. The service failed to start, I checked the logs per the instructions and that is the error that is in the logs. 

Here is the complete output with context and all:
```
siggib@pretalx:~$ sudo systemctl start pretalx-web
siggib@pretalx:~$ sudo systemctl status pretalx-web
× pretalx-web.service - pretalx web service
     Loaded: loaded (/etc/systemd/system/pretalx-web.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Thu 2023-07-06 13:07:17 UTC; 8s ago
    Process: 14169 ExecStart=/var/pretalx/.local/bin/gunicorn pretalx.wsgi --name pretalx --workers 4 --max->   Main PID: 14169 (code=exited, status=200/CHDIR)
        CPU: 5ms

Jul 06 13:07:17 pretalx systemd[1]: pretalx-web.service: Scheduled restart job, restart counter is at 5.
Jul 06 13:07:17 pretalx systemd[1]: Stopped pretalx web service.
Jul 06 13:07:17 pretalx systemd[1]: pretalx-web.service: Start request repeated too quickly.
Jul 06 13:07:17 pretalx systemd[1]: pretalx-web.service: Failed with result 'exit-code'.
Jul 06 13:07:17 pretalx systemd[1]: Failed to start pretalx web service.
siggib@pretalx:~$ sudo journalctl -u pretalx-web
Jul 05 22:12:54 pretalx systemd[1]: Started pretalx web service.
Jul 05 22:12:54 pretalx systemd[4873]: pretalx-web.service: Changing to the requested working directory fail>Jul 05 22:12:54 pretalx systemd[4873]: pretalx-web.service: Failed at step CHDIR spawning /var/pretalx/.loca>Jul 05 22:12:54 pretalx systemd[1]: pretalx-web.service: Main process exited, code=exited, status=200/CHDIR
Jul 05 22:12:54 pretalx systemd[1]: pretalx-web.service: Failed with result 'exit-code'.
Jul 05 22:12:54 pretalx systemd[1]: pretalx-web.service: Scheduled restart job, restart counter is at 1.
Jul 05 22:12:54 pretalx systemd[1]: Stopped pretalx web service.
Jul 05 22:12:54 pretalx systemd[1]: Started pretalx web service.
Jul 05 22:12:54 pretalx systemd[4879]: pretalx-web.service: Changing to the requested working directory fail>Jul 05 22:12:54 pretalx systemd[4879]: pretalx-web.service: Failed at step CHDIR spawning /var/pretalx/.loca>Jul 05 22:12:54 pretalx systemd[1]: pretalx-web.service: Main process exited, code=exited, status=200/CHDIR
Jul 05 22:12:54 pretalx systemd[1]: pretalx-web.service: Failed with result 'exit-code'.
Jul 05 22:12:54 pretalx systemd[1]: pretalx-web.service: Scheduled restart job, restart counter is at 2.
Jul 05 22:12:54 pretalx systemd[1]: Stopped pretalx web service.
Jul 05 22:12:54 pretalx systemd[1]: Started pretalx web service.
Jul 05 22:12:54 pretalx systemd[4880]: pretalx-web.service: Changing to the requested working directory fail>Jul 05 22:12:54 pretalx systemd[4880]: pretalx-web.service: Failed at step CHDIR spawning /var/pretalx/.loca>Jul 05 22:12:54 pretalx systemd[1]: pretalx-web.service: Main process exited, code=exited, status=200/CHDIR
Jul 05 22:12:54 pretalx systemd[1]: pretalx-web.service: Failed with result 'exit-code'.
Jul 05 22:12:55 pretalx systemd[1]: pretalx-web.service: Scheduled restart job, restart counter is at 3.
Jul 05 22:12:55 pretalx systemd[1]: Stopped pretalx web service.
Jul 05 22:12:55 pretalx systemd[1]: Started pretalx web service.
Jul 05 22:12:55 pretalx systemd[4881]: pretalx-web.service: Changing to the requested working directory fail>Jul 05 22:12:55 pretalx systemd[4881]: pretalx-web.service: Failed at step CHDIR spawning /var/pretalx/.loca>Jul 05 22:12:55 pretalx systemd[1]: pretalx-web.service: Main process exited, code=exited, status=200/CHDIR
Jul 05 22:12:55 pretalx systemd[1]: pretalx-web.service: Failed with result 'exit-code'.
Jul 05 22:12:55 pretalx systemd[1]: pretalx-web.service: Scheduled restart job, restart counter is at 4.
Jul 05 22:12:55 pretalx systemd[1]: Stopped pretalx web service.
Jul 05 22:12:55 pretalx systemd[1]: Started pretalx web service.
Jul 05 22:12:55 pretalx systemd[4882]: pretalx-web.service: Changing to the requested working directory fail>Jul 05 22:12:55 pretalx systemd[4882]: pretalx-web.service: Failed at step CHDIR spawning /var/pretalx/.loca>Jul 05 22:12:55 pretalx systemd[1]: pretalx-web.service: Main process exited, code=exited, status=200/CHDIR
Jul 05 22:12:55 pretalx systemd[1]: pretalx-web.service: Failed with result 'exit-code'.
Jul 05 22:12:55 pretalx systemd[1]: pretalx-web.service: Scheduled restart job, restart counter is at 5.
Jul 05 22:12:55 pretalx systemd[1]: Stopped pretalx web service.
Jul 05 22:12:55 pretalx systemd[1]: pretalx-web.service: Start request repeated too quickly.
Jul 05 22:12:55 pretalx systemd[1]: pretalx-web.service: Failed with result 'exit-code'.
Jul 05 22:12:55 pretalx systemd[1]: Failed to start pretalx web service.
Jul 05 23:04:28 pretalx systemd[1]: Started pretalx web service.
Jul 05 23:04:28 pretalx systemd[6900]: pretalx-web.service: Changing to the requested working directory fail>Jul 05 23:04:28 pretalx systemd[6900]: pretalx-web.service: Failed at step CHDIR spawning /var/pretalx/.loca>Jul 05 23:04:28 pretalx systemd[1]: pretalx-web.service: Main process exited, code=exited, status=200/CHDIR
Jul 05 23:04:28 pretalx systemd[1]: pretalx-web.service: Failed with result 'exit-code'.
Jul 05 23:04:28 pretalx systemd[1]: pretalx-web.service: Scheduled restart job, restart counter is at 1.
Jul 05 23:04:28 pretalx systemd[1]: Stopped pretalx web service.
Jul 05 23:04:28 pretalx systemd[1]: Started pretalx web service.
Jul 05 23:04:28 pretalx systemd[6906]: pretalx-web.service: Changing to the requested working directory fail>Jul 05 23:04:28 pretalx systemd[6906]: pretalx-web.service: Failed at step CHDIR spawning /var/pretalx/.loca>Jul 05 23:04:28 pretalx systemd[1]: pretalx-web.service: Main process exited, code=exited, status=200/CHDIR
Jul 05 23:04:28 pretalx systemd[1]: pretalx-web.service: Failed with result 'exit-code'.
Jul 05 23:04:29 pretalx systemd[1]: pretalx-web.service: Scheduled restart job, restart counter is at 2.
Jul 05 23:04:29 pretalx systemd[1]: Stopped pretalx web service.
Jul 05 23:04:29 pretalx systemd[1]: Started pretalx web service.
Jul 05 23:04:29 pretalx systemd[6908]: pretalx-web.service: Changing to the requested working directory fail>
siggib@pretalx:~$ 
```
Here is the systemD file I created per step 6 in the directions:
```
siggib@pretalx:~$ sudo cat /etc/systemd/system/pretalx-web.service
[Unit]
Description=pretalx web service
After=network.target

[Service]
User=pretalx
Group=pretalx
WorkingDirectory=/var/pretalx/.local/lib/python3.8/site-packages/pretalx
ExecStart=/var/pretalx/.local/bin/gunicorn pretalx.wsgi \
                      --name pretalx --workers 4 \
                      --max-requests 1200  --max-requests-jitter 50 \
                      --log-level=info --bind=127.0.0.1:8345
Restart=on-failure

[Install]
WantedBy=multi-user.target
siggib@pretalx:~$ 
```

**Reply by rixx** [reply]
Sorry, I can't help you in that case – your systemd file clearly states that your working directory is meant to be `/var/pretalx/.local/lib/python3.8/site-packages/pretalx`, and that you're *executing* `/var/pretalx/.local/bin/gunicorn`, which is an executable file on your system. If you can both confirm that the gunicorn file is an executable file and that `/var/pretalx/.local/lib/python3.8/site-packages/pretalx` is where your pretalx installation is (did you check that?), then the issue exists outside of pretalx.
